File: ECE39500/homeworks/HW05/MatrixMultiply.java
class MatrixMultiply {

   double a[ ][ ];
   double b[ ][ ];
   double c[ ][ ];
   int cColLB; int cColUB;
   int cRowLB; int cRowUB;

   public MatrixMultiply(double a[ ][ ], double b[ ][ ], double c[ ][ ], 
                        int colsPartitions, int thisColsPartition, 
                        int rowsPartitions, int thisRowsPartition) {

      this.a = a;
      this.b = b;
      this.c = c;
      partition(rowsPartitions, thisRowsPartition, colsPartitions, thisColsPartition, c);
   }

   private int partitionLB(int numPartitions, int thisPartition, int length)  {
      int lb = (thisPartition*length)/numPartitions; 
      return lb;
   }

   private int partitionUB(int numPartitions, int thisPartition, int length) {
      int ub = ((thisPartition+1)*length)/numPartitions - 1;
      return ub;
   }

   private void partition(int rowsPartitions, int thisRowsPartition, 
                         int colsPartitions, int thisColsPartition,
                         double a[ ][ ]) {
      cRowLB = partitionLB(rowsPartitions, thisRowsPartition, a.length);
      cRowUB = partitionUB(rowsPartitions, thisRowsPartition, a.length);
      cColLB = partitionLB(colsPartitions, thisColsPartition, a[0].length);
      cColUB = partitionUB(colsPartitions, thisColsPartition, a[0].length);
   }

   public void mm( ) {
      for (int i = cRowLB; i < cRowUB+1; i++) {
         for (int j = cColLB; j < cColUB+1; j++) {
            int cij = 0;
            for (int k = 0; k < b.length; k++) {
               cij += a[i][k]*b[k][j];
            }
            c[i][j]+= cij;
         }
      }
   }

  public String toString( ) {
     String matrixMultiply = "c[" + c.length + "][" + c[0].length+ "] = " +
                             "a[" + a.length + "][" + a[0].length + "] X " +
                             "b[" + b.length + "][" + cColLB + ":" + cColUB + "]\n";

     matrixMultiply += "c[" + cRowLB + ":"+cRowUB + "][" +cColLB + ":"+ cColUB +" ] = "+
                       "a[" + cRowLB + ":" + cRowUB +"][0:" + b.length +"] X "+
                    "b[0:" + b.length +"][" +cColLB + ":" +cColUB + "]";
      return matrixMultiply;
   }
}
         

File: ECE39500/homeworks/HW05/MatrixMultiply.py
class MatrixMultiply:
    def __init__(self, a, b, c, 
                 colsPartitions, 
                 thisColsPartition, 
                 rowsPartitions, 
                 thisRowsPartition):
        self.a = a
        self.b = b
        self.c = c

        self.partition(rowsPartitions, thisRowsPartition, colsPartitions, thisColsPartition, self.c)
        return

    def partitionLB(self, numPartitions, thisPartition, length):
        return ((thisPartition*length)/numPartitions)
    
    def partitionUB(self, numPartitions, thisPartition, length):
        return (((thisPartition+1)*length)/numPartitions)-1
    
    def partition(self, rowsPartitions, thisRowsPartition,
                  colsPartitions, thisColsPartition, a):
        self.cRowLB = self.partitionLB(rowsPartitions, thisRowsPartition, len(self.c))
        self.cRowUB = self.partitionUB(rowsPartitions, thisRowsPartition, len(self.c))
        self.cColLB = self.partitionLB(colsPartitions, thisColsPartition, len(self.c[0]))
        self.cColUB = self.partitionUB(colsPartitions, thisColsPartition, len(self.c[0]))
        return None
    
    def mm(self):
        for i in range(int(self.cRowLB), int(self.cRowUB+1)):
            for j in range(int(self.cColLB), int(self.cColUB+1)):
                cij = 0
                for k in range(len(self.b)):
                    cij += self.a[i][k]*self.b[k][j]
                self.c[i][j] += cij
        return None

    def __str__(self):
        mmstr = f"""c[{len(self.c)}][{len(self.c[0])}] = a[{len(self.a)}][{len(self.a[0])}] X b[{len(self.b)}][{self.cColLB}:{self.cColUB}]"""
        mmstr += f"""c[{self.cRowLB}:{self.cRowUB}][{self.cColLB}:{self.cColUB}] = a[{self.cRowLB}:{self.cRowUB}][0:{len(self.b)}] X b[0:{len(self.b)}][{self.cColLB}:{self.cColUB}]"""
        return mmstr